webcrypto/subtle_crypto.go

```diff
@@ -137,8 +141,87 @@ func (sc *SubtleCrypto) Verify(
 //   - SHA-512
 //
 // The `data` parameter should contain the data to be digested.
-func (sc *SubtleCrypto) Digest(algorithm goja.Value, data interface{}) *goja.Promise {
-	return nil
+func (sc *SubtleCrypto) Digest(algorithm interface{}, data goja.Value) *goja.Promise {
+	promise, resolve, reject := sc.makeHandledPromise()
+
+	// Validate that the value we received is either an ArrayBuffer, TypedArray, or DataView
+	// This uses the technique described in https://github.com/dop251/goja/issues/379#issuecomment-1164441879
+	var (
+		isArrayBuffer = IsInstanceOf(sc.vu.Runtime(), data, ArrayBufferConstructor)
+		isDataView    = IsInstanceOf(sc.vu.Runtime(), data, DataViewConstructor)
+		isTypedArray  = IsTypedArray(sc.vu.Runtime(), data)
+	)
+
+	if !isArrayBuffer && !isDataView && !isTypedArray {
+		reject(fmt.Errorf("data must be an ArrayBuffer, TypedArray, or DataView"))
+		return promise
+	}
+
+	// 2.
+	// Cast the data to a Goja Object, and, as we're now sure it's
+	// either an ArrayBuffer, or a view on an ArrayBuffer, we can
+	// get the underlying ArrayBuffer by exporting its buffer property
+	// to a Goja ArrayBuffer, and then getting its underlying Go slice
+	// by calling the `Bytes()` method.
+	//
+	// Doing so conviniently also copies the underlying buffer, which
+	// is required by the specification.
+	// See https://www.w3.org/TR/WebCryptoAPI/#SubtleCrypto-method-digest
+	asObject := data.ToObject(sc.vu.Runtime())
+	arrayBuffer, ok := asObject.Export().(goja.ArrayBuffer)
+	if !ok {
+		reject(fmt.Errorf("could not get ArrayBuffer from data"))
+		return promise
+	}
+	bytes := arrayBuffer.Bytes()
+
+	// The specification explicitly requires us to copy the underlying
+	// bytes held by the array buffer
+	bytesCopy := make([]byte, len(bytes))
+	copy(bytesCopy, bytes)
+
+	// 3.
+	normalizedAlgorithm, err := NormalizeAlgorithm(algorithm, OperationIdentifierDigest)
+	if err != nil {
+		// "if an error occurred, return a Promise rejected with NormalizedAlgorithm"
+		reject(err)
+		return promise
+	}
+
+	alg, ok := normalizedAlgorithm.(Algorithm)
```

Comment:
Out of curiosity, why does `NormalizeAlgorithm` returns an interface, not the `Algorithm`?

Reply:
I believe the reason for that will become clearer in the further PRs you look at. In the context of digest not so much 👍🏻 

In short there are many kinds of `Algorithm`. It could be an instance of `KeyAlgorithm`, an `RSAHashedKeyAlgorithm`, or one of a couple of others. And each actions and their own set of data structures will come with different expectations. For instance RSA expects the Algorithm field to hold a `RSAHashedKeyAlgorithm` flavor, which exposes the hash algorithm used.

 I considered using Generics to handle that case, but in my experience at this point with the lib and with dealing with goja, it's easier and more convenient to do runtime checking and casting rather than trying to be too clever 😞

Reply:
That I can understand, but still, it sounds like instead of the null interface, there can be some specific interface that can combine the most generic representation of the algorithm.

As an example bellow could look like this:
```golang
type NormalizedAlgorithm interface {
   Algorithm() string 
}


func NormalizeAlgorithm(algorithm interface{}, op OperationIdentifier) (NormalizedAlgorithm, error) {
 ...
}

switch alg.Algorithm() {}

```
```
Otherwise, whenever you use this method, you should do casting.

Reply:
I believe you're onto something 😄 I need to think about it a bit more, and I'll get back to this comment.

In the meantime, here's how `NormalizeAlgorithm` is used in a more "complex" and "standard" use case. Digest is a limited and quite specific subset indeed.

Reply:
I've substantially cleaned the algorithm normalization up. Plus, I added a `NormalizedAlgorithm` interface which allows doing exactly what you requested. I'm not sure it will survive the "fire baptism" in a larger context (`generateKey`, `encrypt`, etc), but the design is already better like this, and it will force us to find better designs down the road indeed 🙇🏻 👍🏻